ipaserver/plugins/whoami.py

```diff
@@ -0,0 +1,141 @@
+request. The information includes:
+
+ * type of object
+ * command to retrieve details of the object
+ * arguments and options to pass to the command
+
+EXAMPLES:
+
+ Look up as IPA user:
+   kinit admin
+   ipa whoami
+   ------------------------------------------
+   object: user
+   details: user_show/1
+   arguments: admin
+   options:
+   ------------------------------------------
+
+ Look up as a user from a trusted domain:
+   kinit user@AD.DOMAIN
+   ipa whoami
+   ------------------------------------------
+   object: idoverrideuser
+   details: idoverrideuser_show/1
+   arguments: ('default trust view', 'user@ad.domain')
+   options:
+   ------------------------------------------
+
+ Look up as a host:
+   kinit -k
+   ipa whoami
+   ------------------------------------------
+   object: host
+   details: host_show/1
+   arguments: ipa.example.com
+   options:
+   ------------------------------------------
+
+ Look up as a Kerberos service:
+   kinit -k -t /path/to/keytab HTTP/ipa.example.com
+   ipa whoami
+   ------------------------------------------
+   object: service
+   details: service_show/1
+   arguments: HTTP/ipa.example.com
+   options:
+   ------------------------------------------
+""")
+
+register = Registry()
+
+
+@register()
+class whoami(Command):
+    __doc__ = _('Describe currently authenticated identity.')
+
+    output_params = (
+        Str('object', label=_('Object class name')),
+        Str('details', label= _('Function to get details')),
+        Str('arguments*', label=_('Arguments to details function')),
+        Str('options*', label=_('Options to details function')),
+    )
+
+    has_output = (
+        output.Output('object', unicode, _('Object class name')),
+        output.Output('details', unicode, _('Function to get details')),
+        output.Output('arguments', list, _('Arguments to details function')),
+        output.Output('options', list, _('Options to details function')),
+    )
+
+    def execute(self, **options):
+        """
+        Retrieve the DN we are authenticated as to LDAP and find bindable IPA
+        object that handles the container where this DN belongs to. Then report
+        details about this object.
+        """
+        exceptions = dict(
+            idoverrideuser=(DN("cn={0}".format(DEFAULT_TRUST_VIEW_NAME)),
+                            DEFAULT_TRUST_VIEW_NAME, 'ipaOriginalUid'),
+        )
+        ldap = api.Backend.ldap2
+
+        # whoami_s() call returns a string 'dn: <actual DN value>'
+        # We also reject ldapi-as-root connections as DM is a virtual object
+        dn = DN(ldap.conn.whoami_s()[4:])
+        if dn == DN('cn=Directory Manager'):
+            raise errors.NotFound(
+                    reason=_('Cannot query Directory Manager with API'))
+
+        entry = ldap.get_entry(dn)
+        o_name = None
+        o_func = None
+        o_args = []
+        o_opts = []
+        for o in api.Object():
```

Comment:
There is no need to init it because api is fully initialized already at the point when any plugin command is run. api.Object() returns an iterator, it is *not* a class constructor.